cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -54,20 +60,33 @@ protected AbstractProgramRuntimeService(ProgramRunnerFactory programRunnerFactor
   public synchronized RuntimeInfo run(Program program, ProgramOptions options) {
     ProgramRunner runner = programRunnerFactory.create(ProgramRunnerFactory.Type.valueOf(program.getType().name()));
     Preconditions.checkNotNull(runner, "Fail to get ProgramRunner for type " + program.getType());
-    final SimpleRuntimeInfo runtimeInfo = new SimpleRuntimeInfo(runner.run(program, options), program);
-    addRemover(runtimeInfo);
+    ProgramOptions optionsWithRunId = addRunId(options, RunIds.generate());
+    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithRunId), program);
+    programStarted(runtimeInfo);
     runtimeInfos.put(runtimeInfo.getType(), runtimeInfo.getController().getRunId(), runtimeInfo);
     return runtimeInfo;
   }
 
+  private ProgramOptions addRunId(ProgramOptions options, RunId runId) {
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    builder.put(ProgramOptionConstants.RUN_ID, runId.getId());
+
+    return new SimpleProgramOptions(options.getName(), new BasicArguments(builder.build()), options.getUserArguments(),
+                                    options.isDebug());
+  }
+
+  protected RuntimeInfo createRuntimeInfo(ProgramController controller, Program program) {
+    return new SimpleRuntimeInfo(controller, program);
+  }
+
+  protected List<RuntimeInfo> getRuntimeInfos() {
```

Comment:
I believe this needs to be synchronized since it access the `runtimeInfos` table.